### PR TITLE
Bug: Unexpected behavior when assigning multi-dimensional array to DataFrame column 

### LINKED ISSUE #40827: BUG: Unexpected behavior when assigning multi-dimensional array to DataFrame column
- [x] I have searched the [[pandas] tag](https://stackoverflow.com/questions/tagged/pandas) on StackOverflow for similar questions.

- [x] I have asked my usage related question on [StackOverflow](https://stackoverflow.com).

---

#### Question about pandas

My question is essentially the same as the StackOverflow question [“Unexpected behavior in assigning 2d numpy array to pandas DataFrame”](https://stackoverflow.com/q/52142475), which, although it has an accepted answer, states in an edit to the question text:

> Since the bounty is going to expiry [sic], I accepted the most voted answer. But it does not provide a answer to the above questions.

I agree, and still don’t understand the behavior described in the question, which I will restate here:
Given the following code:
```
import pandas
import numpy
df = pandas.DataFrame(numpy.zeros((4, 1)), columns=['A'])
y = numpy.random.randn(4, 2)
df['A'] = y
```
which results in (e.g.)
```
df
#           A
# 0  0.994449
# 1 -0.280376
# 2 -0.993288
# 3  0.009625
y
# array([[ 0.99444899, -0.80571928],
#        [-0.28037615, -0.47071158],
#        [-0.99328776, -0.15674664],
#        [ 0.00962502,  1.24896221]])
```

> I expect it to throw an exception because of shape mismatch. But pandas silently accepted the assignment: y's first column is assigned to df.
> 
> Is this an intentional design? If yes, what is the rationale behind?

### PR DESCRIPTION
**Pull request by weikhor**
- [ ] closes #40827
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/internals/managers.py [hunk 1]
```diff
@@ -1057,6 +1057,9 @@ def iset(
                     "Shape of new values must be compatible with manager shape"
                 )
 
+            if value.shape[0] > self.shape[0]:
+                raise AssertionError
```
Comment (phofl) [inline pandas/core/internals/managers.py:1061]: This needa a better error message

File: pandas/tests/series/indexing/test_setitem.py [hunk 2]
```diff
@@ -1665,3 +1665,14 @@ def test_setitem_empty_mask_dont_upcast_dt64():
     ser.mask(mask, "foo", inplace=True)
     assert ser.dtype == dti.dtype  # no-op -> dont upcast
     tm.assert_series_equal(ser, orig)
+
+
+def test_setitem_multi_dimension_array_to_dataframe():
+    # GH#40827
+    df = DataFrame(np.zeros((4, 1)), columns=["A"])
+    y = np.random.randn(4, 3)
```
Comment (phofl) [inline pandas/tests/series/indexing/test_setitem.py:1673]: What happens here, if the column A has duplicates?

Comment (weikhor) [reply pandas/tests/series/indexing/test_setitem.py:1673]: Sorry I do not fully understand this. 
`column A has duplicates`
Are you mean that column A has duplicated values like this?

df
#           A
# 0  1.2
# 1 1.2
# 2 1.2
# 3  1.2

Comment (phofl) [reply pandas/tests/series/indexing/test_setitem.py:1673]: Sorry bad wording, I meant a df with duplicated column names

File: pandas/core/internals/managers.py (hunk 1 above)
Comment (WillAyd) [reply pandas/core/internals/managers.py:1061]: Wondering if we should just be mirroring what numpy does here?

```python
>>> arr = np.zeros((4, 1))
>>> arr2 = np.random.randn(4, 2)
>>> arr[:, 0] = arr2
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: could not broadcast input array from shape (4,2) into shape (4,)
```

Somewhat surprised this doesn't happen natrually

File: pandas/tests/series/indexing/test_setitem.py (hunk 2 above)
Comment (weikhor) [reply pandas/tests/series/indexing/test_setitem.py]: If df with duplicated column names, with version 1.5.0.dev0+1217.g6e1a040142
```
df = pd.DataFrame(numpy.zeros((4, 2)), columns=['A', 'A'])
y = numpy.random.randn(4, 3)

df['A'] = y

print(df)
```
Result
```
          A         A
0 -0.256262  0.197093
1 -0.628350 -0.111466
2  0.394070  0.111357
3  0.780353 -0.20774
```

Comment (phofl) [reply pandas/tests/series/indexing/test_setitem.py]: The interesting part is when the rhs has as many columns as A is duplicated

**Review by phofl — COMMENTED** [review]
This needs more tests:

- loc
- iloc
- regular setitem
- isetitem

Then we need scalar indexers, list-like indexers. Also we should add tests for a rhs with an Index like a DataFrame and an array like you have right now

**Comment by weikhor** [comment]
I close this since I stuck this for long time. I get simpler task. Thank for review.

File: pandas/tests/series/indexing/test_setitem.py (hunk 2 above)
Comment (weikhor) [reply pandas/tests/series/indexing/test_setitem.py]: When the rhs has as many columns as A is duplicated,
```
df = DataFrame(
    [[1, 2, 3, 4], [1, 2, 3, 4]],
    columns=["A", "A", "A", "A"],
)
df['A'] = np.array([[5, 6, 7, 8], [5, 6, 7, 8]])
print(df)
```
Result:
```
   A  A  A  A
0  5  6  7  8
1  5  6  7  8

```

**Comment by weikhor** [comment]
I will close this first. Need some time to understand.  Thank for review

**Pull request closed without merging**